xdsl/dialects/func.py

```diff
@@ -47,6 +56,97 @@ def verify_(self) -> None:
                 "input types"
             )
 
+    @classmethod
+    def parse(cls, parser: Parser) -> FuncOp:
+        # Parse visibility keyword if present
+        visibility = parser.tokenizer.next_token_of_pattern(
+            re.compile("public|nested|private")
+        )
+        if isinstance(visibility, Span):
+            visibility = visibility.text
+
+        # Parse function name
+        name = parser.parse_symbol_name().data
+
+        # Parse function arguments
+        args = parser.parse_comma_separated_list(
+            parser.Delimiter.PAREN,
+            lambda: parser.parse_optional_argument()
+            or parser.parse_optional_type()
+            or parser.raise_error("Expected argument or type"),
```

Comment:
more shenanigans!

Reply:
Mh, here I would agree it helps readability significantly though.
I also find your other suggestion (`if self.function_type.outputs:`) kind of contradictory TBH.
But this discussion already happened, I'll switch to a heavier style!

Reply:
I agree that it's more concise, but easy on the eyes is not easy on the brain :) There is a very big difference between this and the other thing, in that `None` treated as falsey is very dangerous, because if your type implements truthyness (like ArrayAttr might one day, seen as it's not also an iterator), an empty array type might be treated as falsey, returning an error. This would be surprising to me, at least

Reply:
Here the [diff](https://github.com/xdslproject/xdsl/pull/992/commits/cefad81f486d07626e5c60c19500eaa18f17cb0b#diff-cc6fedbf5a9cce29a2bac9e03bdc84ec6aca641ea9313c2aba1779c437d21ba2) for reference.

Reply:
Didn't read your answer before mine, sorry about that!
Good point for the ArrayAttr, I agree 👍